src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```diff
@@ -118,12 +192,203 @@
     return rankings;
   }
 
-  // parse the given file into a List of CastVoteRecords for tabulation
+  // helper to extract the required CVRContest from the input CVR
+  CVRContest getCvrContest(CVR cvr, Contest contestToTabulate) {
+    CVRContest cvrContestToTabulate = null;
+    // find current snapshot
+    CVRSnapshot currentSnapshot = null;
+    if (cvr.CVRSnapshot != null) {
+      for (CVRSnapshot cvrSnapshot : cvr.CVRSnapshot) {
+        if (cvrSnapshot.ObjectId.equals(cvr.CurrentSnapshotId)) {
+          currentSnapshot = cvrSnapshot;
+          break;
+        }
+      }
+    }
+    // find CVRContest which matches the Contest we are tabulating
+    if (currentSnapshot.CVRContest != null) {
+      for (CVRContest cvrContest : currentSnapshot.CVRContest) {
+        if (cvrContest.ContestId.equals(contestToTabulate.ObjectId)) {
+          cvrContestToTabulate = cvrContest;
+          break;
+        }
+      }
+    }
+    return cvrContestToTabulate;
+  }
+
+  void parseXml(List<CastVoteRecord> castVoteRecords) {
+    try {
+
+      // load XML
+      XmlMapper xmlMapper = new XmlMapper();
+      xmlMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+      FileInputStream inputStream = new FileInputStream(new File(filePath));
+      CastVoteRecordReport cvrReport = xmlMapper.readValue(inputStream, CastVoteRecordReport.class);
+      if (cvrReport.Election.length > 1) {
+        Logger.warning("Multiple Election objects found.  Only the first one will be processed.");
+      }
+
+      // Find the contest for tabulation
+      Contest contestToTabulate = null;
+      for (Contest contest : cvrReport.Election[0].Contest) {
+        if (contest.Name.equals(this.contestId)) {
+          contestToTabulate = contest;
+          break;
+        }
+      }
+      if (contestToTabulate == null) {
+        Logger.severe("Contest \"%s\" from config file not found!", this.contestId);
```

Comment:
`Logger.severe` (and likely the other new log functions) are not WAI. When I tested triggering this message, it said:

> "SEVERE: Contest "%s" from config file not found!"

Reply:
Good catch.